themes/default/content/blog/converting-full-terraform-programs-to-pulumi/index.md

```diff
@@ -0,0 +1,524 @@
+
+First, clone the repo and `cd` into the directory containing the Terraform project:
+
+```bash
+$ git clone https://github.com/avantoss/vault-infra.git
+$ cd vault-infra/terraform/main
+```
+
+Next, run the converter:
+
+{{< chooser language "typescript,python,go,csharp" >}}
+{{% choosable language typescript %}}
+
+```bash
+$ pulumi convert --from terraform --language typescript --out pulumi
+```
+
+{{% /choosable %}}
+{{% choosable language python %}}
+
+```bash
+$ pulumi convert --from terraform --language python --out pulumi
+```
+
+{{% /choosable %}}
+{{% choosable language go %}}
+
+```bash
+$ pulumi convert --from terraform --language go --out pulumi
+```
+
+{{% /choosable %}}
+{{% choosable language csharp %}}
+
+```bash
+$ pulumi convert --from terraform --language csharp --out pulumi
+```
+
+{{% /choosable %}}
+{{< /chooser >}}
+
+![pulumi convert](pulumi-convert-from-terraform.gif)
+
+The converted code is generated in the specified `pulumi` output directory. There are two primary files generated from this program's conversion:
+
+* {{< langfile >}} contains the converted code for the main program
+* <pulumi-chooser type="language" options="typescript,python,go,csharp" option-style="none" class="inline">
+    <pulumi-choosable type="language" value="typescript"><code>vault.ts</code></pulumi-choosable>
+    <pulumi-choosable type="language" value="python"><code>vault.py</code></pulumi-choosable>
+    <pulumi-choosable type="language" value="go"><code>vault.go</code></pulumi-choosable>
+    <pulumi-choosable type="language" value="csharp"><code>vault.cs</code></pulumi-choosable>
+  </pulumi-chooser> contains the <code>Vault</code> Pulumi component, converted from the Terraform module
+
+### Addressing TODOs
+
+The file for the `Vault` component makes up the bulk of the implementation and contains some TODOs emitted by the converter that must be addressed manually.
+
+#### Replace
+
+For example, the converter doesn't yet support the `replace` built-in function.
+
+{{< chooser language "typescript,python,go,csharp" >}}
+{{% choosable language typescript %}}
+
+```typescript
+name: notImplemented("replace(var.name_prefix,\"_\",\"-\")"),
+```
+
+{{% /choosable %}}
+{{% choosable language python %}}
+
+```python
+name=not_implemented("replace(var.name_prefix,\"_\",\"-\")"),
+```
+
+{{% /choosable %}}
+{{% choosable language go %}}
+
+```go
+Name: notImplemented("replace(var.name_prefix,\"_\",\"-\")"),
+```
+
+{{% /choosable %}}
+{{% choosable language csharp %}}
+
+```csharp
+Name = NotImplemented("replace(var.name_prefix,\"_\",\"-\")"),
+```
+
+{{% /choosable %}}
+{{< /chooser >}}
+
+We can fill in an implementation.
```

Comment:
```suggestion
We can fill in an implementation.  Note that we get to use the full expressiveness of the native and familiar string manipulation libraries in our target programming language, instead of the relatively constrained options of the Terraform DSL.
```